Repo URL - https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

Commit ID - 54bc995ab4151135d98f7e6eab3f063ba71d991b

In [2]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split

## Part 1. Loading the dataset

In [3]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
data = pd.read_csv('science_data_large.csv')
print(data.head(15))
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
label_column = 'Size nm^3'
X = data.drop(columns=[label_column])
y = data[label_column]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)

## Part 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set
from sklearn.linear_model import LinearRegression

# Create a linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = pd.DataFrame({
    'Temperature °C': [665],
    'Mols KCL': [110]
})
predicted_output = model.predict(sample_data)
print(f"Predicted Size nm^3: {predicted_output[0]}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f"Model score (R^2): {score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

Predicted Size nm^3: 280192.3426223573
Model score (R^2): 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


The model score $R^2$ for our model is 0.85525. This shows the "goodness of fit" of the model by measuring the proportion of the variance in the dependent variable as $R^2 = 1 − \frac{\text{Sum squared regression}}{\text{Total sum of squares}}$. Furthermore this means that since the R^2 scoes is closer to 1 that means it accounts for more of the variance, and the data points are close to the regression line.

Equation: $$h(x) = 866.14641 \cdot \text{Temperature °C } + 1032.69507 \cdot \text{Mols KCL} - 409391.47958$$

## Part 4. Use Cross Validation

In [6]:
from sklearn.model_selection import cross_val_score
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X, y, cv=5)

# Report on their finding and their significance
print(f"Cross-validation scores: {scores}")
print(f"Mean cross-validation score: {scores.mean()}")

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.8568167899144437


The significance of these cross-validation scores is that each score in the array represents the performance of the model on one of the folds during the cross-validation process. In this case, we have 5 scores, each corresponding to a different split of the dataset. We see that the mean of these scors is 0.85682, meaning that our $R^2$ value is very similar to our one from part 3, which tells us that our regression model is fitting the data consistently across different subsets of the dataset.

## Part 5. Using Polynomial Regression

In [7]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score

# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Split the augmented dataset into training and testing sets (90% train, 10% test)
X_train_poly, X_test_poly, y_train, y_test = train_test_split(X_poly, y, test_size=0.10, random_state=8)
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

# Evaluating the model with metrics
r2_train_score = poly_model.score(X_train_poly, y_train)
print(f"R^2 Score of Trained Polynomial Regression: {r2_train_score}")
r2_test_score = poly_model.score(X_test_poly, y_test)
print(f"R^2 Score of Test Polynomial Regression: {r2_test_score}")

# Extract coefficients and intercept for the polynomial equation

# Extract coefficients and intercept
coefficients_poly = poly_model.coef_
intercept_poly = poly_model.intercept_
feature_names = poly.get_feature_names_out(input_features=X.columns)
   

# Assuming feature_names is defined as per the previous context
for name, coeff in zip(feature_names, coefficients_poly):
    print(f"{name}: {coeff}")

print(f"\nIntercept: {intercept_poly}")

R^2 Score of Trained Polynomial Regression: 1.0
R^2 Score of Test Polynomial Regression: 1.0
1: 0.0
Temperature °C: 12.000000028954805
Mols KCL: -1.271076937548951e-07
Temperature °C^2: -3.921130087292113e-11
Temperature °C Mols KCL: 2.000000000012829
Mols KCL^2: 0.02857142870560825

Intercept: 1.4160992577672005e-05


Equation: 

$h(x) = 1.416 \times 10^{-5} + 12.0000 \cdot \text{Temperature °C} + (-1.27 \times 10^{-7}) \cdot \text{Mols KCL} + (-3.92) \times 10^{-12} \cdot (\text{Temperature °C})^2 + 2.0000 \cdot (\text{Temperature °C} \cdot \text{Mols KCL}) + 0.0285 \cdot (\text{Mols KCL})^2$

Without outputs outside 5 sig figs:

$h(x) = 1.416 \times 10^{-5} + 12 \cdot \text{Temperature °C} + 2 \cdot (\text{Temperature °C} \cdot \text{Mols KCL}) + 0.0285 \cdot (\text{Mols KCL})^2$

Description:
The $R^2$ value is now 1.0 which means that we fit the data perfectly, and the "goodness of fit" is perfect. We account for each data point perfectly in both the trained and test sets. Since it fits like this we now have the equation of a slime and can predict the output, meaning we save we world.